vslib/src/SwitchStateBase.cpp

```diff
@@ -2424,3 +2441,191 @@ void SwitchStateBase::debugSetStats(
+                voq_switch_id = (int32_t) attr_list[i].value.u32;
+
+                if (voq_switch_id < 0)
+                {
+                    SWSS_LOG_ERROR("Invalid VOQ switch id %d", voq_switch_id);
+                    return (int32_t)SAI_STATUS_INVALID_ATTR_VALUE_0 + (int32_t)i;
+                }
+                break;
+
+            case SAI_SWITCH_ATTR_MAX_SYSTEM_CORES:
+                voq_max_cores = attr_list[i].value.u32;
+
+                if (voq_max_cores < 1)
+                {
+                    SWSS_LOG_ERROR("Invalid VOQ max system cores %d", voq_max_cores);
+                    return (int32_t)SAI_STATUS_INVALID_ATTR_VALUE_0 + (int32_t)i;
+                }
+                break;
+
+            case SAI_SWITCH_ATTR_SYSTEM_PORT_CONFIG_LIST:
+                sys_port_count = attr_list[i].value.sysportconfiglist.count;
+                sys_port_cfg_list = attr_list[i].value.sysportconfiglist.list;
+
+                if (sys_port_count < 1 || !sys_port_cfg_list)
+                {
+                    SWSS_LOG_ERROR("Invalid voq system port config info! sys port count %d, sys port list %p", sys_port_count, sys_port_cfg_list);
+                    return (int32_t)SAI_STATUS_INVALID_ATTR_VALUE_0 + (int32_t)i;
+                }
+                break;
+
+            default:
+                // Ignore other attributes. They are processed elsewhere.
+                break;
+        }
+    }
+
+    if (!voq_switch)
+    {
+        // No switch type attribute in the attribute list.
+        return SAI_STATUS_SUCCESS;
+    }
+
+    CHECK_STATUS(create_system_ports(voq_switch_id, sys_port_count, sys_port_cfg_list));
+
+    CHECK_STATUS(set_system_port_list());
+
+    CHECK_STATUS(set_voq_switch_attributes(voq_switch_id, voq_max_cores));
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::create_system_ports(
+       _In_ int32_t voq_switch_id,
+       _In_ uint32_t sys_port_count,
+       _In_ sai_system_port_config_t *sys_port_cfg_list)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_INFO("create system ports");
+
+    m_system_port_list.clear();
+
+    for (uint32_t i = 0; i < sys_port_count; i++)
+    {
+        SWSS_LOG_DEBUG("create system port index %u", i);
+
+        sai_object_id_t system_port_id;
+
+        CHECK_STATUS(create(SAI_OBJECT_TYPE_SYSTEM_PORT, &system_port_id, m_switch_id, 0, NULL));
+
+        m_system_port_list.push_back(system_port_id);
+
+        sai_attribute_t attr;
+
+        attr.id = SAI_SYSTEM_PORT_ATTR_CONFIG_INFO;
+        attr.value.sysportconfig = sys_port_cfg_list[i];
+
+        CHECK_STATUS(set(SAI_OBJECT_TYPE_SYSTEM_PORT, system_port_id, &attr));
+
+        attr.id = SAI_SYSTEM_PORT_ATTR_TYPE;
+        attr.value.s32 = SAI_SYSTEM_PORT_TYPE_REMOTE;
+
+        if ((int32_t)sys_port_cfg_list[i].attached_switch_id == voq_switch_id)
+        {
+            attr.value.s32 = SAI_SYSTEM_PORT_TYPE_LOCAL;
+
+            // Need to set local port oid attribute TODO
```

Comment:
what local port ?

Reply:
System port config list passed at the time system initialization is the system port config of all ports in entire voq system. If any given system port config corresponds to local port (identified by switch_id), the attribute SAI_SYSTEM_PORT_ATTR_TYPE should be SAI_SYSTEM_PORT_TYPE_LOCAL when read. Also, if the type is local, the corresponding local port id should be available in SAI_SYSTEM_PORT_ATTR_PORT. To get local port oid from m_port_list corresponding to the system port, the core and core port index values are used. This processing is in PR #703 (vslib/src/SwitchStateBase.cpp: line number 2557 - 2594)

Reply:
from your description seems like some system ports may correspond to local port and some not is that correct ? what is the purpose of such configuration ? Is there ad design spec somewhere about system ports ?

Reply:
Please refer to the VOQ HLD: https://github.com/Azure/SONiC/blob/master/doc/voq/voq_hld.md
If the system has only one asic, the number of system ports will be same as number of local ports. We have just system ports configured for the local ports only. We still need the mapping between local port and system port as needed by the portsorch to work (PR: https://github.com/Azure/sonic-swss/pull/1431) . If we have multi asic/multi line card system, System port config will have list of system ports all asics from all line cards. This mapping is need by PortsOrch to differentiate between local ports and system Ports initialization in portsorch.

Reply:
i will dive to this document and see what i can understand. for now it seems to me like voq switch is another switch inside current switch, and it will be really hard to track all dependencies between those 2 internal switches and monitor all read only variables to be refreshed in correct way.
**_also currently when switch is created only 3 other objects are created internally: queues, ingress_priority_gorups and scheduler_groups. and all those 3 objects are considered dependencies of each port. and there are special task taken inside sairedis and syncd to process those dependencies especially on port add and port remove.
this voq system is adding another dependency which is system_port, and when this will be in use, another special handling  for this needs to be added in syncd/sairedis, this will be essential in warm boot scenario to not lost any objects,
also i dont know what will/should happen to voq system ports that have related local ports in case when local port will be removed  _**

Reply:
also, currently sonic is using only 1 asic, but its written in way to support multiple asic/switches, but that path is not tested in any way yet, so i would not be surprised if there are many bugs related to multi asic sonic at current stage, not even mentioning support for multi asic voq

Reply:
As per current design,  for the PR: Azure/sonic-swss#1431 system port action for port removal does not exist. For now, to get 1431 through, I'll associate local port to system ports using info from current m_port_list. When the remove is handled in system port, we would consider an appropriate refresh model for system ports.

Reply:
yes, system port remove/create dont exits, but local port remove/create exists.
Also is this VOQ switch supported on Broadcom or Mellanox chip ? or this is on some new asic/silocon ?

Reply:
As far as I know, currently VOQ switch support is available for Broadcom chip (DNX/Jerico). No idea about Mellanox and other chipsets.

Reply:
> what local port ? what is relation between system_port and port ?

SystemPort is applicable in VOQ based chassis system. When we say "system_port" it either port in local asic or port in remote asic on the same chassis. Every local port(Front panel port) will system port defined in VOQ chassis system. 

This doesnt apply any system that aren't VOQ.

Reply:
> yes, system port remove/create dont exits, but local port remove/create exists.
> Also is this VOQ switch supported on Broadcom or Mellanox chip ? or this is on some new asic/silocon ?

VOQ switch is supported by broadcom currently in SONiC. I dont think/know, Mellanox has VOQ based chip. 

This work is for new chipset which are based on VOQ. (first one to onboard to SONiC is broadcom DNX chipset).